nexus/types/src/deployment/clickhouse.rs

```diff
@@ -0,0 +1,118 @@
+    pub cluster_name: String,
+    /// An arbitrary string shared by all nodes used at runtime to determine
+    /// whether nodes are part of the same cluster.
+    pub cluster_secret: String,
+
+    /// This is used as a marker to tell if the raft configuration
+    /// in a new inventory collection is newer than the last collection.
+    //
+    /// This serves as a surrogate for the log index of the last committed
+    /// configuration, which clickhouse keeper doesn't expose.
+    ///
+    /// This is necesssary because during inventory collection we poll
+    /// multiple keeper nodes, and each returns their local knowledge of the
+    /// configuration. But we may reach different nodes in different attempts,
+    /// and some nodes in a following attempt may reflect stale configuration.
+    /// Due to timing, we can always query old information. That is just normal
+    /// polling. However, we never want to use old configuration if we have
+    /// already seen and acted on newer configuration.
+    pub highest_seen_keeper_leader_committed_log_index: u64,
+
+    /// The desired state of the clickhouse keeper cluster
+    ///
+    /// We decouple deployment of zones that should contain clickhouse keeper
+    /// processes from actually starting or stopping those processes, adding or
+    /// removing them to/from the keeper cluster, and reconfiguring other keeper
+    /// and clickhouse server nodes to reflect the new configuration.
+    ///
+    /// As part of this decoupling, we keep track of the intended zone
+    /// deployment in the blueprint, but that is not enough to track the desired
+    /// state of the keeper cluster. We are only allowed to add or remove one
+    /// keeper node at a time, and therefore we must track the desired state of
+    /// the keeper cluster which may change multiple times until the keepers in
+    /// the cluster match the deployed zones. An example may help:
+    ///
+    ///   1. We start with 3 keeper nodes in 3 deployed keeper zones and need to
+    ///      add two to reach our desired policy of 5 keepers
+    ///   2. The planner adds 2 new keeper zones to the blueprint
+    ///   3. The planner will also add **one** new keeper process that matches
+    ///      one of the deployed zones to the desired keeper cluster.
+    ///   4. The executor will start the new keeper process, attempt to add it
+    ///      to the keeper cluster by pushing configuration updates to the other
+    ///      keepers, and then updating the clickhouse server configurations to
+    ///      know about the new keeper.
+    ///   5. If the keeper is successfully added, as reflected in inventory, then
+    ///      steps 3 and 4 above will be retried for the next keeper process.
+    ///   6. If the keeper is not successfully added by the executor it will
+    ///      continue to retry indefinitely.
+    ///   7. If the zone is expunged while the planner is has it as part of its
```

Comment:
`is has it as part of its desired state`? 😄

Reply:
Fixed.